### PR TITLE
Fix borderRadius when `double.infinity`

### PR DESCRIPTION
**Pull request by bernaferrari**
Fix issue where the output of `BorderRadius.circular(double.infinity)` is a rectangle, not a stadium/circle.

This is an experiment. I want to see how many tests break. Another approach might be using int.maxInt, maxFloat or just `radius == double.infinity, then act`.

cc @dnfield

### CODE REVIEW

File: packages/flutter/lib/src/painting/border_radius.dart [hunk 1]
```diff
@@ -398,12 +400,13 @@ class BorderRadius extends BorderRadiusGeometry {
     // Because the current radii could be negative, we must clamp them before
     // converting them to an RRect to be rendered, since negative radii on
     // RRects don't make sense.
+    final Radius maxRadius = Radius.circular(math.max(rect.width, rect.height));
```
Comment (dnfield) [inline packages/flutter/lib/src/painting/border_radius.dart:403]: Is this right? If the user actually specified an elliptical radius we'll potentially convert it to circular here.

I think this should probably be
```suggestion
    final Radius maxRadius = Radius.elliptical(rect.width, rect.height);
```

File: packages/flutter/test/painting/border_radius_test.dart [hunk 2]
```diff
@@ -187,8 +187,8 @@ void main() {
     expect(borderRadius.resolve(TextDirection.ltr).toRRect(rect), RRect.fromRectXY(rect, 3.0, 3.0));
     expect(borderRadius.resolve(TextDirection.rtl).toRRect(rect), RRect.fromRectXY(rect, 3.0, 3.0));
 
-    const Radius radius1 = Radius.elliptical(89.0, 87.0);
-    const Radius radius2 = Radius.elliptical(103.0, 107.0);
+    const Radius radius1 = Radius.elliptical(8.9, 8.7);
+    const Radius radius2 = Radius.elliptical(9.3, 9.7);
```
Comment (dnfield) [inline packages/flutter/test/painting/border_radius_test.dart:191]: We should also add tests that assert a radius larger than the rect gets clamped.

I can help you running this through the Google monorepo as well.

File: packages/flutter/lib/src/painting/border_radius.dart (hunk 1 above)
Comment (dnfield) [reply packages/flutter/lib/src/painting/border_radius.dart:403]: Also, the comment above should get updated to explain this.

**Comment by dnfield** [comment]
Kicked of a global presubmit at http://test/OCL:518142794:BASE:518142911:1679365815777:a1cd727a (googlers only, sorry)

File: packages/flutter/test/painting/border_radius_test.dart (hunk 2 above)
Comment (bernaferrari) [reply packages/flutter/test/painting/border_radius_test.dart:191]: Yes, I need to make a suite of tests. I just wanted to see if it would break something first.

File: packages/flutter/lib/src/painting/border_radius.dart (hunk 1 above)
Comment (bernaferrari) [reply packages/flutter/lib/src/painting/border_radius.dart:403]: I'm going to tell you a secret: 
image

But what you suggested makes more sense than what I did. I'll try that. Thanks!

**Comment by bernaferrari** [comment]
Even if tests pass, I found an edge scenario:
https://user-images.githubusercontent.com/351125/226509930-0074796b-41b5-43ce-95fc-2caa8ffabc92.mov

I guess the maxInt/maxFloat idea is better? :P

File: packages/flutter/test/painting/border_radius_test.dart (hunk 2 above)
Comment (bernaferrari) [reply packages/flutter/test/painting/border_radius_test.dart]: Added them.

**Comment by dnfield** [comment]
maxInt/maxFloat definitely won't work. They'll get truncated to infinity and dropped.

I'm not sure I understand what I'm seeing in the video.

**Comment by bernaferrari** [comment]
The width was 200, and if I set the value to 400, 800, 1000, 10000, it still made visible changes..

**Comment by bernaferrari** [comment]
I have 2 suggestions:
- We set 3.402823e+38 (max float value) as the upper bounds, that would fix `double.infinity` problem.
- We check if each Radius is circular (if so, set to width/height). We need to figure out the logic for elliptical).

Right now, I commited the first approach. It works here. Let's see if tests will pass.

**Comment by dnfield** [comment]
It'd be really nice if we can figure out the elliptical case and not need to have a special constant floating around the framework for this.

If we really need the special constant, I think it should be exposed from `dart:ui` as something like `kMaxScalar` or something.

**Comment by bernaferrari** [comment]
I'll take a look another day. Possibly diving one by another and checking if the result is small enough could be a strategy, but it would be a) slower than 3.402823e+38 and b) still would fail, if say, someone had a rectangle with a width of 3.40282**4**e+38 (which I guess is really unusual).

**Comment by bernaferrari** [comment]
### GPT-4

> Consider the following in Flutter.
> 1. For w: 20, h:20, borderRadius.circular(20) is acceptable.
> 2. For w: 20, h:20, borderRadius.circular(40) is not acceptable, can be reduced to borderRadius.circular(20).
>
> The same can't be said for BorderRadius.elliptical.
> For w: 400, h: 50, Radius.elliptical(100000000, 1000) still shows a value that can be seen visually.
>
> Generalize a math formula that allows to find the "minimum necessary Radius for a given width, height, radiusX and radiusY".

Let w and h be the width and height of a rectangle, respectively. Let r_x and r_y be the horizontal and vertical radii of an elliptical border, respectively. Our goal is to find the minimum necessary radius r_min_x and r_min_y for a given width, height, r_x, and r_y.

We know, by definition, that the elliptical border must satisfy the equation:

(x/w)^2 + (y/h)^2 = 1

The top-right corner of the rectangle is at the point (w/2, h/2). After substituting these values into the equation, we get:

((w/2)/w)^2 + ((h/2)/h)^2 = 1
(1/2)^2 + (1/2)^2 = 1
1/4 + 1/4 = 1
1/2 = 1

Now, we will scale the radii to find the minimum necessary radii r_min_x and r_min_y that satisfy the equation:

(r_min_x/w)^2 + (r_min_y/h)^2 = 1

Since we have already established that 1/2 is the maximum value for both terms on the left-hand side of this equation, we can set r_min_x and r_min_y as follows:

r_min_x = w / (2 * sqrt(2))
r_min_y = h / (2 * sqrt(2))

Consequently, if given r_x and r_y values are greater than these r_min_x and r_min_y values, we can use these new minimum radii to create a visually acceptable elliptical border. If r_x or r_y is smaller than these minimum radii, you should use the provided values. So, given any r_x and r_y:

r_final_x = min(r_x, w / (2 * sqrt(2)))
r_final_y = min(r_y, h / (2 * sqrt(2)))

In summary, we can use the above formulas to find the minimum necessary radii r_final_x and r_final_y for a given width, height, r_x, and r_y.

**Comment by bernaferrari** [comment]
I just spent some time playing with it. It is hard. If you keep pushing BorderRadius.ellipse, things do get suuuuuper weird (which is probably unintended). So how do we find a balance? I don't know.

The same thing happens if you paint a border of 200 in a widget of size 20, the border gets suuuuper weird. So, there is a max limit, which we need to add (soon). But for sizing, well, it is much more complex.
Specially given IMO (I'm no math person), borders interact with each other, so topLeft radius interact with bottomLeft corner. I don't think I am able to find a way to do that...

But if you find the people that work on Skia/Impeller in Google's corridor, you can ask them how they do it and how we can find that value.

There is an issue about max int, max things, but it's been quiet for a while :(
https://github.com/dart-lang/sdk/issues/41717

**Comment by bernaferrari** [comment]
lhrn just replied:

> The don't necessarily need to be in the platform libraries. It's just constants, anyone can define them.

**Comment by dnfield** [comment]
Defining the max for a scalar in the framework isn't quite right, becuase it's actually the engine that defines the size of the scalar type.

**Comment by bernaferrari** [comment]
Ops, wrong button. The benefit of the framework is that we can play with impeller or Skia in the future, maybe impeller fixes this issue, maybe not, so we have power to define what we do with the constant and which constant we use. In the engine might be weird..

But if we wanted in the engine, who should we talk with?

**Comment by bernaferrari** [comment]
A new day, a new proposal. We could bound when BorderRadius.circular and not bound when Elliptical (for now). This would solve 99% of the issues with minimal math, no math.sqrt or math.pow. This is the best balance we've gotten so far. 4 if's, 1 math.min. Not a constant. Not a complex operation either.

File: packages/flutter/lib/src/painting/border_radius.dart [hunk 3]
```diff
@@ -398,12 +400,23 @@ class BorderRadius extends BorderRadiusGeometry {
     // Because the current radii could be negative, we must clamp them before
     // converting them to an RRect to be rendered, since negative radii on
     // RRects don't make sense.
+    //
+    // We also clamp the radii to be no larger than half the width or height of
+    // the rect, since radii larger than that don't make sense either.
+    // We only do this if the radii are circular, since non-circular radii
+    // can be larger than half the width or height of the rect.
```
Comment (dnfield) [inline packages/flutter/lib/src/painting/border_radius.dart:407]: Please file a bug to figure out what to do for non-circular radii and add a TODO linking to it here. Ideally it should have details from what you've attempted so far and what works/doesn't work.

**Review by dnfield — DISMISSED** [review]
This LGTM with a nit about adding a TODO

**Comment by Piinks** [comment]
(Triage), there are many failing checks here. @bernaferrari this is approved, but it looks like @dnfield requested a TODO be added?

**Comment by bernaferrari** [comment]
I added the documentation but forgot to commit the link. The issue will be Google Testing. I'll try to make the change a bit "less agressive" and hope for the best, but I guess I'll still need the help from one of you (if you think this change is worthy). Let's see in the next few days.

**Comment by bernaferrari** [comment]
I think the only way to land this without triggering half the Google Testing would be with the constant (would only affect double.infinity and nothing else) :(

**Comment by dnfield** [comment]
The customer test failures look real, are they?

**Comment by bernaferrari** [comment]
They are new, I think `super_editor` wasn't in customer testing before this PR and now it is. So when we initially tried, only Google Testing failed, but now Google Testing + super_editor.

**Comment by dnfield** [comment]
It looks like macos_ui is failing too. At least those give you some open source bugs to work with :)

**Comment by bernaferrari** [comment]
I could fix those, but at the end there will still be google testing :(

I changed from min(width, height) / 2 to clamp(min(width, height)/2) to see if the google testing goldens don't fail (the screenshot you sent me before).

How do you feel? Should we try to fix everything? There isn't an easy path. Maybe I should add an intermediate method, migrate them to that, migrate back to the old method, etc? Isn't just using the float32 value easier/better?

**Comment by dnfield** [comment]
I think the customer_testing failures are the same kind of issue as what I'ms eeing in the google-internal testing. It's ok if we don't fix this right now. 

If we want to use the float32 value, we should just make sure to expose it from dart:ui rather than having a random constant for it floating in the framework - it's dart:ui that defines the max value for a scalar, and that could change in the future.

**Comment by bernaferrari** [comment]
There are two kind of fails, the "standard" (value changed) and the golden (the one you shared, probably due to rounding errors). Hopefully the second is no more. But it will still be a pain to rewrite everything.

Thanks, I'll open an issue there.

**Comment by goderbauer** [comment]
(Triage) @bernaferrari What is the status here? Do you still have plans for this PR?

**Comment by bernaferrari** [comment]
I don't know the plan, I opened a PR on the Engine to update the comment (so that we can figure a solution from the engine side), I got without time and things got frozen. But this specific PR is kind of dead?

**Comment by goderbauer** [comment]
> But this specific PR is kind of dead?

Ok, let's close it then.
